cli: Support clean up for `persist-nic-names` command

cathay4t opened the pull request:
The whole workflow:

 * Use `nmstatectl persist-nic-names` to pin __every__ ethernet interface to its
   MAC via systemd .link file.
 * After boot into new environment, `nmstatectl persist-nic-names --cleanup`
   will remove un-required systemd .link file.

Key tech point:

    Use `udevadm test-builtin net_id /sys/class/net/<iface_name>` to check
    systemd preferred interface name and compare to current interface
    name.

Extra changes:

 * Removed `should_persist_nic_for()` as we are pinning every ethernet
   interface.
 * Comment changed from `\\\` to `\\` as this is not public
   documentation.
 * The `--cleanup` might require chroot if running environment holds
   different systemd/udev config than real host.

Tested:
 * CentOS stream 9 VM by renaming a existing interface to new name.

 * RHEL 8.8 and CentOS stream 9 on Mellanox MT27710:
    * Not from upgrade, just reinstall CentOS stream 9 and restored the
      nmstate created .link files.

cathay4t commented:
Test logs in CentOS stream 9 VM after renaming `ens10` to `ens9` using `ip link` command.

[fge@c9s ~]$ sudo ncl persist-nic-names
[2023-04-11T05:51:04Z INFO  nmstatectl::persist_nic] Would persisting the interface ens9 with MAC 52:54:00:42:03:95
[2023-04-11T05:51:04Z INFO  nmstatectl::persist_nic] Systemd network link file created at /etc/systemd/network/98-nmstate-ens9.link
[2023-04-11T05:51:04Z INFO  nmstatectl::persist_nic] Would persisting the interface ens3 with MAC 52:54:00:23:b7:dc
[2023-04-11T05:51:04Z INFO  nmstatectl::persist_nic] Systemd network link file created at /etc/systemd/network/98-nmstate-ens3.link

[fge@c9s ~]$ sudo ncl persist-nic-names --cleanup
[2023-04-11T05:51:09Z INFO  nmstatectl::persist_nic] Found persisted NIC(ens9) file: 98-nmstate-ens9.link
[2023-04-11T05:51:09Z INFO  nmstatectl::persist_nic] Found persisted NIC(ens3) file: 98-nmstate-ens3.link
[2023-04-11T05:51:09Z INFO  nmstatectl::persist_nic] The generated /etc/systemd/network/98-nmstate-ens3.link file has no effect for interface ens3, removing
[2023-04-11T05:51:09Z INFO  nmstatectl::persist_nic] Systemd generate interface name 'ens10' != pinned name 'ens9', will keep config file /etc/systemd/network/98-nmstate-ens9.link

cathay4t commented:
Test log on Mellanox MT27710:
 * RHEL 8

[root@hpe-apache-cn99xx-03 nmstate]# nmstatectl  persist-nic-names
[2023-04-11T05:32:24Z INFO  nmstatectl::persist_nic] Would persisting the interface enp11s0f0 with MAC f4:03:43:fe:60:f6
[2023-04-11T05:32:24Z INFO  nmstatectl::persist_nic] Systemd network link file created at /etc/systemd/network/98-nmstate-enp11s0f0.link
[2023-04-11T05:32:24Z INFO  nmstatectl::persist_nic] Would persisting the interface enp11s0f1 with MAC f4:03:43:fe:60:f7
[2023-04-11T05:32:24Z INFO  nmstatectl::persist_nic] Systemd network link file created at /etc/systemd/network/98-nmstate-enp11s0f1.link

 * CentOS stream 9:

[root@hpe-apache-cn99xx-03 ~]# nmstatectl persist-nic-names --cleanup --dry-run
[2023-04-11T08:12:20Z INFO  nmstatectl::persist_nic] Found persisted NIC(enp11s0f0) file: 98-nmstate-enp11s0f0.link
[2023-04-11T08:12:20Z INFO  nmstatectl::persist_nic] Found persisted NIC(enp11s0f1) file: 98-nmstate-enp11s0f1.link
[2023-04-11T08:12:20Z INFO  nmstatectl::persist_nic] Systemd generate interface name 'enp11s0f1np1' != pinned name 'enp11s0f1', will keep config file /etc/systemd/network/98-nmstate-enp11s0f1.link
[2023-04-11T08:12:20Z INFO  nmstatectl::persist_nic] Systemd generate interface name 'enp11s0f0np0' != pinned name 'enp11s0f0', will keep config file /etc/systemd/network/98-nmstate-enp11s0f0.link

cgwalters commented on rust/src/cli/persist_nic.rs:
```diff
@@ -195,49 +117,152 @@ pub(crate) fn run_persist_immediately(
+    link_dir.join(format!("{PERSIST_FILE_PREFIX}-{iface_name}.link"))
+}
+
+fn extract_iface_names_from_link_file(file_name: &str) -> Option<String> {
+    if file_name.ends_with(".link")
+        && file_name.starts_with(PERSIST_FILE_PREFIX)
+    {
+        Some(
+            file_name[(PERSIST_FILE_PREFIX.len() + 1)
+                ..(file_name.len() - ".link".len())]
+                .to_string(),
+        )
+    } else {
+        None
+    }
+}
```
There's an off by one error here, see
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=047d386fc9c3d1bd460e3173915d3678

This function would be a great candidate for unit tests which would have caught this.

But see my combinator-based version which fixes this bug and is easier to read and should also avoid any runtime array length checks.

cathay4t replied:
Nice one! Thanks for the suggest!

cgwalters commented on rust/src/cli/persist_nic.rs:
```diff
@@ -195,49 +117,152 @@ pub(crate) fn run_persist_immediately(
 
-    Ok("".to_string())
+    for line in output.split('\n') {
+        if line.starts_with(&format!("{}=", ID_NET_NAME_ONBOARD)) {
```
Should also use `strip_prefix()` here.

cgwalters commented on rust/src/cli/persist_nic.rs:
```diff
@@ -166,23 +81,30 @@ pub(crate) fn run_persist_immediately(
         log::info!(
-            "{action} the interface with MAC {mac} to \
-                        interface name {iface_name}"
+            "Would persisting the interface {} with MAC {mac}",
```
s/Would persisting/Would persist/

cgwalters commented on rust/src/cli/persist_nic.rs:
```diff
@@ -52,108 +60,15 @@ fn gather_state() -> Result<NetworkState, CliError> {
     Ok(state)
 }
 
-fn device_is_virtual(iface: &Interface) -> bool {
```
Don't we still want this?  On an OCP node I didn't want to persist any of the dynamic ovs interfaces I think...

cathay4t replied:
OVS internal interface is not ethernet, so I think we are good here. For OVS system interface, I think the OVS  database still want the interface name been persistent.

cgwalters commented on rust/src/cli/persist_nic.rs:
```diff
@@ -1,47 +1,55 @@
 
 use crate::error::CliError;
 
-/// Comment added into our generated link files
+// Comment added into our generated link files
```
`///` is standard for documenting specific items, even though this isn't a public API I think it's best practice to do so.

cathay4t replied:
Sure. Will revert it back to `///`.

cgwalters reviewed (commented):
A first quick pass

cgwalters commented:
> The --cleanup might require chroot if running environment holds
different systemd/udev config than real host.

Yes, this is currently going to be required; in the OCP flow we're shipping this nmstate binary in a privileged container which is based on rhel8.

cathay4t commented:
@cgwalters New patch uploaded. Thanks for the review!

cgwalters commented on rust/src/cli/persist_nic.rs:
```diff
@@ -195,49 +116,146 @@ pub(crate) fn run_persist_immediately(
     })?;
 
-    Ok("".to_string())
+    for line in output.split('\n') {
+        if let Some(n) = line.strip_prefix(&format!("{}=", ID_NET_NAME_ONBOARD))
+        {
+            return Ok(n.to_string());
+        }
+    }
+
+    for line in output.split('\n') {
+        if let Some(n) = line.strip_prefix(&format!("{}=", ID_NET_NAME_SLOT)) {
+            return Ok(n.to_string());
+        }
+    }
+
+    for line in output.split('\n') {
+        if let Some(n) = line.strip_prefix(&format!("{}=", ID_NET_NAME_PATH)) {
+            return Ok(n.to_string());
+        }
+    }
```
Performance doesn't matter much here but I think this loop would be more runtime efficient (iterating *once* over the input and not allocating a string in each iteration) and actually easier to read like this:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9da17e9abc0d767f61d983c5007c724c

cgwalters replied:
Ah sorry, it's more subtle than that, but I typed up https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=45c60b4a9c8300812e1ee851ad5e5d0c

cathay4t replied:
the `fold` function cause the iterator goes through the whole output, while my code return the value once ideal value found. So the performance difference need to debate. 

And the `fold` code is not rust-learner friendly. It looks more like perl to me.

cgwalters replied:
It's a bit subtle, because in practice we can only skip iterating the whole output if we find `ID_NET_NAME_ONBOARD`.  If we find `ID_NET_NAME_SLOT` we still need to finish in case `ID_NET_NAME_ONBOARD` will come after.

I don't have a strong opinion here; but the string-allocation in a loop seems unnecessary.
How about https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7ba83c816fabe298655742c9766122b3 ?

cathay4t replied:
Adopted the new `split_once()` code.

cgwalters commented on rust/src/cli/persist_nic.rs:
```diff
@@ -195,49 +116,146 @@ pub(crate) fn run_persist_immediately(
-            "NOTE: would persist the interface with MAC {mac} to interface name {iface_name}"
-        );
-        Ok(())
+    for (iface_name, file_path) in pinned_ifaces.drain() {
```
In most cases `into_iter()` is preferred over `drain` because it can be more efficient to also de-allocate the backing container at the same time.  It's actually the default for `for` loops https://doc.rust-lang.org/std/iter/index.html#for-loops-and-intoiterator so we can just drop the `.drain()` here.

cathay4t replied:
Thanks for the hint. Fixed.

cgwalters commented on rust/src/cli/persist_nic.rs:
```diff
@@ -195,49 +116,146 @@ pub(crate) fn run_persist_immediately(
+            continue;
+        }
+        let systemd_iface_name =
+            match get_systemd_preferred_iface_name(&iface_name) {
```
Still need to pass down `root` here

cgwalters commented on rust/src/cli/persist_nic.rs:
```diff
@@ -195,49 +116,146 @@ pub(crate) fn run_persist_immediately(
+    iface_name: &str,
+) -> Result<String, CliError> {
+    let output: String = String::from_utf8(
+        std::process::Command::new("udevadm")
```
And if we have a root, use `chroot`.

cathay4t replied:
This `chroot /new_root udevadm` require `sysfs` been mounted in new root, is that true for openshift use case?

cgwalters replied:
Yep, we are (recursively) bind mounting the host `/`, so we'll see everything the host does.

cathay4t committed:
cli: Support clean up for `persist-nic-names` command

The whole workflow:

 * Use `nmstatectl persist-nic-names` to pin *every* ethernet interface to its
   MAC via systemd .link file.
 * After boot into new environment, `nmstatectl persist-nic-names --cleanup`
   will remove un-required systemd .link file.

Key tech point:

    Use `udevadm test-builtin net_id /sys/class/net/<iface_name>` to check
    systemd preferred interface name and compare to current interface
    name.

Extra changes:

 * Removed `should_persist_nic_for()` as we are pinning every ethernet
   interface.
 * Comment changed from `\\\` to `\\` as this is not public
   documentation.
 * The `--cleanup` might require chroot if running environment holds
   different systemd/udev config than real host.

Tested:
 * CentOS stream 9 VM by renaming a existing interface to new name.

 * RHEL 8.8 and CentOS stream 9 on Mellanox MT27710:
    * Not from upgrade, just reinstall CentOS stream 9 and restored the
      nmstate created .link files.

Signed-off-by: Gris Ge <fge@redhat.com>

cgwalters commented:
Looking forward to this!  One thing that'd probably be good as a followup is some sort of automated CI for this.

cathay4t commented:
> 

Our current CI setup is running tests inside of a container where there is no real ethernet interface to play with.
I am afraid we need downstream QE on CI automation like we did for tests on SR-IOV, Infiniband and etc.

cathay4t commented:
I have uploaded RHEL 8(also works in RHEL 9) compiled binary for this PR:

https://people.redhat.com/fge/bz_2182769/nmstatectl_persist_nic_names_2

jlebon commented:
Should `--cleanup` delete the stamp file too? One could imagine wanting to do this again in the future (e.g. for RHEL9 to RHEL10) and it'd make sense to reuse the same stamp file instead of having different ones each time since nmstate itself is agnostic to the RHEL version transition happening. But also from the MCO point of view, being able to key off of the presence of the stamp file to know if to run `--cleanup` would be helpful so that we don't have to maintain our own secondary stamp.

jlebon commented:
Threw something up in https://github.com/nmstate/nmstate/pull/2316! Let's chat there.
